fuel/transformers/__init__.py

```diff
@@ -130,45 +183,112 @@ class SingleMapping(Transformer):
+
+        """
+        raise NotImplementedError(
+            "`{}` does not support examples as input, but the wrapped data "
+            "stream produces examples.".format(self.__class__.__name__))
+
+    def transform_source_batch(self, source_batch):
+        """Applies a transformation to a batch from a source.
+
+        Parameters
+        ----------
+        source_batch : :class:`numpy.ndarray`
+            A batch of examples from a source.
+
+        """
+        raise NotImplementedError(
+            "`{}` does not support batches as input, but the wrapped data "
+            "stream produces batches.".format(self.__class__.__name__))
+
+    def transform_example(self, example):
+        return self._apply_sourcewise_transformation(
+            data=example, method=self.transform_source_example)
+
+    def transform_batch(self, batch):
+        return self._apply_sourcewise_transformation(
+            data=batch, method=self.transform_source_batch)
+
+
+@add_metaclass(ABCMeta)
+class AgnosticSourcewiseTransformer(AgnosticTransformer,
+                                    SourcewiseTransformer):
+    """A sourcewise transformer that operates the same on examples or batches.
+
+    Subclasses must implement the `transform_any_source` method, which is
+    to be applied to both examples and batches. This is useful when the
+    example and batch implementation of a sourcewise transformation are
+    the same.
+
+    """
+    def transform_any(self, data):
+        return self._apply_sourcewise_transformation(
+            data=data, method=self.transform_any_source)
 
     @abstractmethod
-    def mapping(self, source):
-        """Applies a single mapping to selected sources.
+    def transform_any_source(self, source_data):
+        """Applies a transformation to a source.
+
+        The data can either be an example or a batch of examples.
 
         Parameters
         ----------
-        source : :class:`numpy.ndarray`
-            Input source.
+        source_data : :class:`numpy.ndarray`
+            Data from a source.
 
         """
 
-    def get_data(self, request=None):
-        if request is not None:
-            raise ValueError
-        data = list(next(self.child_epoch_iterator))
-        for i, source_name in enumerate(self.data_stream.sources):
-            if source_name in self.which_sources:
-                data[i] = self.mapping(data[i])
-        return tuple(data)
 
+class Flatten(SourcewiseTransformer):
+    """Flattens selected sources.
+
+    If the wrapped data stream produces batches, they will be flattened
+    along all but the first axis.
 
-class Flatten(SingleMapping):
-    """Flattens selected sources along all but the first axis."""
+    Incoming sources will be treated as numpy arrays (i.e. using
+    `numpy.asarray`).
+
+    """
     def __init__(self, data_stream, **kwargs):
-        super(Flatten, self).__init__(data_stream, **kwargs)
+        super(Flatten, self).__init__(
```

Comment:
I vaguely remember @bartvm advocating for an empty string instead of `'features'`, but I don't remember why.

I'm okay with `'features'`, personally.